File: it.stefano.bls.mock/lib/src/main/java/control/BLController.java
package control;

import interf.ILed;
import observable.GenericObserver;

public class BLController extends GenericObserver implements IBLController {
	
	private String curState = "s0";
	private ILed led;
	
	public BLController(ILed led) {
		setLed(led);
	}
	
	@Override
	public void update(String data) {
		if(data.equals("pressed"))
			onPressed();
		else if(data.equals("released"))
			onReleased();
	}
	
	private void onPressed() {
		if(curState.equals("s0")) {
			turnOnLed();
			setState("sOn0");
		} else if(curState.equals("sOn1")) {
			turnOffLed();
			setState("sOff0");
		} else if(curState.equals("sOff1")) {
			turnOnLed();
			setState("sOn0");
		}
	}
	private void onReleased(){
		if(curState.equals("sOn0")) {
			setState("sOn1");
		} else if(curState.equals("sOff0")) {
			setState("sOff1");
		} 
	}
	private void setState(String state) {
		this.curState = state;
	}
	private void turnOffLed() {
		led.off();
	}
	private void turnOnLed() {
		led.on();
	}
	
	private void setLed(ILed aLed) {
		this.led = aLed;
	}
}

File: it.stefano.bls.arduino/ButtonLedSystemArduino/BLController.cpp
#include "BLController.h"

String BLController::fsm(String input) {
  String cmd = "";
  if(input.equals("pressed")) {
    if(state.equals("sOff")) {
      cmd = "on";
      state = "sOn";
    } else if(state.equals("sOn")) {
      cmd = "off";
      state = "sOff";
    }
  }
  return cmd;
}
void BLController::elab(String cmd) {
  if(cmd.equals("on")) {
    turnOnLed();
  } else if(cmd.equals("off")) {
    turnOffLed();
  }
}
void BLController::turnOnLed() {
  led->on();
}
void BLController::turnOffLed() {
  led->off();
}

BLController::BLController(Led* led) {
  this->led = led;
  this->state = "sOff";
}
void BLController::update(String data) {
  //Serial.println("BLController | update " +data);
  String cmd = fsm(data);
  elab(cmd);
}